cli_helper/kart.c

```diff
@@ -209,6 +209,21 @@ void exit_on_alarm(int sig)
     exit(exit_code);
 }
 
+char *socket_filename;
+
+/**
+ * @brief Exit signal handler for SIGINT
+ */
+void exit_on_sigint(int sig)
+{
+    putchar('\n');
+    const char* kill_cmd_template = "lsof -t '%s' | xargs kill";
```

Comment:
kill sends TERM by default, you probably want `-s INT`

I'd probably prefer a non-subprocess implementation - `kill(pid, SIGINT)` is easy enough - are there any C replacements for lsof?

Reply:
AFAICT `lsof` isn't part of most Linux base images. `netstat -p` might be an option on Linux but not macOS, but again `netstat` isn't always installed with the base images.

(We could add a dependency to lsof/netstat/whatever on Linux I guess)

Reply:
Re: replacing lsof with a non-subprocess - no-extra-dependency solution:
lsof looks complicated - when I went looking for C alternatives, people suggested that you just search through the data in /proc until you find what you are looking for, which is what lsof does. Eg here:
https://unix.stackexchange.com/questions/302152/is-there-a-way-to-identify-the-pid-or-cgroup-of-a-socket-without-iterating-thr

Reply:
Re: sending out-of-band requests over the socket
It's possible. Peter actually sends very little over the socket. Return exit codes are not sent over the socket, they are sent using`kill` - so we only ever send messages over the socket in one direction. Here's the current flow:
User runs kart.c
Kart.c finds the socket
Kart.c sends a message containing its pid, FDs, environment, command-line arguments...
Kart.c then just sleeps for up to 24 hours.
helper.py is in a never ending while loop. It receives the message and invokes the appropriate click command.
if the command returns or throws an exception, helper.py sends kill SIGALRM to the calling PID

So with current flow, the helper.py while loop is single-threaded: it listens for a message, then executes the message, etc. I could try and make it multi-threaded so it can listen for a signal to kill itself:
- pro - one less dependency (lsof)
- con - maybe a bit less robust to rely on a process that's already not doing what the user wants, to be listening for signals and kill itself. We only get one chance to kill it - if kart.c is killed and helper.py survives, we end up in a bad state where helper.py is running in the background but connected to your tty, and no ^C special keypresses will do anything to it.

Reply:
Is it an option to call setpgid in the per-invocation execution fork?

eg:
1. Background Kart process is PID 100
2. `kart do stuff` → invokes C helper as PID 200
3. 100 connects to background helper, sends args/file-descriptors
4. 100 forks to 201, calls `exec()`
5. 201 calls `setpgid(100)`**
6. if the user hits ctrl-c, the progress group receives the signal: both 100 & 201 & any further subprocesses, but not 200

** it should probably be the result of `getpgrp()` rather than it's own PID.

Reply:
This looks nicer than lsof - I'll try to get it to work